File: source/_posts/Algorithm/1.leetcode/2_sum/2.两数相加.go
package main
import "fmt"
/*
 * @lc app=leetcode.cn id=2 lang=golang
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 // 定义类
type ListNode struct {
	Val int
	Next *ListNode
}
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil && l2 == nil {
		return nil
	}
	if l1 == nil {
		return l2
	}
	if l2 == nil {
		return l1
	}

	sum := l1.Val + l2.Val
	nextNode := addTwoNumbers(l1.Next, l2.Next)
	if sum < 10 {
		return &ListNode{ Val: sum, Next: nextNode }
	} else {
		tempNode := &ListNode{
			Val: 1,
			Next: nil,
		}
		return &ListNode{
			Val: sum - 10,
			Next: addTwoNumbers(nextNode, tempNode),
		}
	}
}
// @lc code=end



func main() {
	var l1  = &ListNode{2,&ListNode{4,&ListNode{3,nil}}}
	l2  := &ListNode{5,&ListNode{6,&ListNode{4,nil}}}
	addTwoNumbers(l1,l2)
	fmt.Print("Hello from goProject.")
}

File: source/_posts/Algorithm/1.leetcode/Easy/2_sum/2.两数相加.java
/*
 * @lc app=leetcode.cn id=2 lang=java
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if(l2==null) return l1;
        ListNode ret = new ListNode(0);  
        ListNode cur = ret;  
  
        int sum = 0;  
        while (true) {  
            if (l1 != null) {  
                sum += l1.val;  
                l1 = l1.next;  
            }  
            if (l2 != null) {  
                sum += l2.val;  
                l2 = l2.next;  
            }  
            cur.val = sum % 10;  
            sum /= 10;  
            if (l1 != null || l2 != null || sum != 0) {  
                cur = (cur.next = new ListNode(0));  
            } else {  
                break;  
            }  
        }  
        return ret;  
    }
}
// @lc code=end

